includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php

```diff
@@ -685,12 +687,44 @@ static function ( $validation_error ) {
 				 * @var DOMAttr $attr_node
 				 * @var string  $error_code
 				 */
-				list( $attr_node, $error_code ) = $disallowed_attribute;
-				$validation_error['code']       = $error_code;
+				list( $attr_node, $error_code, $error_data ) = $disallowed_attribute;
+				$validation_error['code']                    = $error_code;
 
-				$attr_spec = isset( $merged_attr_spec_list[ $attr_node->nodeName ] ) ? $merged_attr_spec_list[ $attr_node->nodeName ] : [];
-				if ( $this->remove_invalid_attribute( $node, $attr_node, $validation_error, $attr_spec ) ) {
-					$removed_attributes[] = $attr_node;
+				if ( self::DISALLOWED_PROPERTY_IN_ATTR_VALUE === $error_code ) {
+					$properties         = $error_data['original_properties'];
+					$invalid_properties = $error_data['invalid_properties'];
+
+					foreach ( $invalid_properties as $invalid_property ) {
+						$validation_error['attr_property_name'] = $invalid_property;
+						if ( $this->should_sanitize_validation_error( $validation_error, [ 'node' => $attr_node ] ) ) {
+							unset( $properties[ $invalid_property ] );
+						}
+					}
+
+					$valid_properties = array_reduce(
+						array_keys( $properties ),
+						static function ( $attr_value, $property ) use ( $properties ) {
+							if ( ! empty( $attr_value ) ) {
+								$attr_value .= ',';
+							}
+
+							return $attr_value . $property . '=' . $properties[ $property ];
+						}
+					);
+
+					$node->setAttribute( $attr_node->nodeName, $valid_properties );
+				} else {
+					if ( self::MISSING_MANDATORY_PROPERTY === $error_code ) {
+						$validation_error['attr_property_name'] = $error_data['property'];
+					} elseif ( self::MISSING_REQUIRED_PROPERTY_VALUE === $error_code ) {
+						$validation_error['attr_property_name']  = $error_data['property'];
+						$validation_error['attr_property_value'] = $error_data['value'];
```

Comment:
I chose to prefix `attr_` because `property_name` and `property_value` were already in use in the style sanitizer. I suppose renaming those instances to `css_property_*` would resolve that to resolve any ambiguity.

Reply:
Renamed `property_*` to `css_property_*` and `attr_property_*` to `property_*` in https://github.com/ampproject/amp-wp/pull/4137/commits/73d04ff259a2ee23f13cbc749a3a49829255fe59.